processing/src/main/java/org/apache/druid/segment/virtual/ExpressionSelectors.java

```diff
@@ -514,15 +509,45 @@ public void inspectRuntimeShape(RuntimeShapeInspector inspector)
   /**
    * Selectors are not consistent in treatment of null, [], and [null], so coerce [] to [null]
    */
-  private static Object coerceListDimToStringArray(List val)
+  // suppressed because calling toArray creates Object[] instead of Long[] which makes ExprEval.bestEffortOf sad
+  @SuppressWarnings("SimplifyStreamApiCallChains")
+  public static Object coerceListToArray(List val)
   {
-    Object[] arrayVal = val.stream().map(x -> x != null ? x.toString() : x).toArray(String[]::new);
-    if (arrayVal.length > 0) {
-      return arrayVal;
+    if (val != null && val.size() > 0) {
+      Object firstElement = val.get(0);
```

Comment:
Expressions support lists that contain nulls, right? What happens if the first element is null? Also, what happens if the first element is an Integer but later ones are Doubles? i.e. JSON `[1, 2.0]`.

It may be better to examine all elements using some kind of binary type conversion rules (like long + int = long, etc).

Reply:
Updated to examine list elements and follow java auto conversion rules